src/analyses/constant_propagator.cpp

```diff
@@ -278,30 +278,53 @@ bool constant_propagator_domaint::ai_simplify(
 
 bool constant_propagator_domaint::valuest::is_constant(const exprt &expr) const
 {
-  if(expr.id()==ID_side_effect &&
-     to_side_effect_expr(expr).get_statement()==ID_nondet)
-    return false;
+  if(expr.is_constant())
+    return true;
 
-  if(expr.id()==ID_side_effect &&
-     to_side_effect_expr(expr).get_statement()==ID_allocate)
-    return false;
+  if(expr.id() == ID_address_of)
+  {
+    const address_of_exprt &address_of_expr = to_address_of_expr(expr);
 
-  if(expr.id()==ID_symbol)
-    if(replace_const.expr_map.find(to_symbol_expr(expr).get_identifier())==
-       replace_const.expr_map.end())
-      return false;
+    return is_constant_address_of(address_of_expr.object());
+  }
+  else if(expr.id() == ID_typecast)
+  {
+    const typecast_exprt &typecast_expr = to_typecast_expr(expr);
 
-  if(expr.id()==ID_index)
-    return false;
+    return is_constant(typecast_expr.op());
+  }
+  else if(expr.id() == ID_array_of)
+  {
+    return is_constant(expr.op0());
+  }
+  else if(
+    expr.id() == ID_plus || expr.id() == ID_array || expr.id() == ID_struct ||
+    expr.id() == ID_union)
+  {
+    forall_operands(it, expr)
```

Comment:
I thought `forall_operands` was now deprecated? Could use:

```C++
return std::all_of(
  expr.operands().begin(), 
  expr.operands().end(), 
  [](const exprt &) { return is_constant(expr); });
```
To make the condition more explicit